File: node/src/sensor/occupancy.ts
import { Sensor, SensorType, SensorCategory } from './base.js';
import { defineSensor, SensorDomain } from './meta.js';

import type { Observable } from '../observable/index.js';
import type { PropertyChangeOf, SensorLike, SensorOptions } from './base.js';

/**
 * Property names of an occupancy sensor.
 *
 * @internal
 */
export enum OccupancyProperty {
  /** Whether occupancy is detected (true = occupied). */
  Detected = 'detected',
}

/**
 * Property values of an occupancy sensor.
 *
 * @internal
 */
export interface OccupancySensorProperties {
  [OccupancyProperty.Detected]: boolean;
}

/** Read-only proxy interface for an occupancy sensor. */
export interface OccupancySensorLike extends SensorLike {
  readonly type: SensorType.Occupancy;
  readonly onPropertyChanged: Observable<PropertyChangeOf<OccupancySensorProperties>>;

  getValue(property: OccupancyProperty.Detected): boolean | undefined;
  getValue(property: string): unknown;
}

/** Occupancy sensor for detecting presence in a room or area. */
export class OccupancySensor<TStorage extends object = Record<string, any>> extends Sensor<OccupancySensorProperties, TStorage> {
  readonly type = SensorType.Occupancy;
  readonly category = SensorCategory.Sensor;

  constructor(name = 'Occupancy Sensor', options?: SensorOptions) {
    super(name, options);

    this._writeState({ [OccupancyProperty.Detected]: false });
  }

  get detected(): boolean {
    return this.props.detected;
  }

  /**
   * Report occupancy state.
   *
   * @param value - True when the area is currently occupied.
   *
   * @example
   * ```ts
   * occupancy.setDetected(true);
   * ```
   */
  setDetected(value: boolean): void {
    this._writeState({ [OccupancyProperty.Detected]: value });
  }

  /**
   * Read-only sensor: external writes are ignored.
   *
   * @internal
   */
  updateValue(_property: string, _value: unknown): void {}
}

/** Registry metadata for {@link OccupancySensor}. */
export const occupancyMeta = defineSensor({
  type: SensorType.Occupancy,
  category: SensorCategory.Sensor,
  assignmentKey: 'occupancy',
  multiProvider: true,
  isDetectionType: false,
  properties: {
    [OccupancyProperty.Detected]: { type: 'boolean', writable: true },
  },
  shortcutable: true,
  cascadeTrigger: { property: OccupancyProperty.Detected, value: true, sustained: true },
  virtual: { properties: { [OccupancyProperty.Detected]: false } },
  semantics: {
    domain: SensorDomain.Binary,
    stateProperty: OccupancyProperty.Detected,
    commandProperty: OccupancyProperty.Detected,
    deviceClass: 'occupancy',
  },
});

File: python/camera_ui_sdk/sensor/occupancy.py
from __future__ import annotations

from collections.abc import Mapping
from enum import Enum
from typing import Any, Generic, Literal, Protocol, overload, runtime_checkable

from typing_extensions import TypedDict, TypeVar

from ..observable import Observable
from .base import Sensor, SensorCategory, SensorLike, SensorType


class OccupancyProperty(str, Enum):
    """Properties for occupancy sensors."""

    Detected = "detected"


class OccupancySensorProperties(TypedDict):
    """Property value map for occupancy sensors."""

    detected: bool


class OccupancyPropertyChangeData(TypedDict):
    """Emitted on OccupancySensorLike.onPropertyChanged."""

    property: str  # OccupancyProperty value
    value: bool


TStorage = TypeVar("TStorage", bound=Mapping[str, Any], default=dict[str, Any])


@runtime_checkable
class OccupancySensorLike(SensorLike, Protocol):
    """Read-only proxy interface for an occupancy sensor."""

    @property
    def type(self) -> SensorType:
        return SensorType.Occupancy

    @overload
    def getValue(self, property: Literal[OccupancyProperty.Detected]) -> bool | None: ...
    @overload
    def getValue(self, property: str) -> object | None: ...

    @property
    def onPropertyChanged(self) -> Observable[OccupancyPropertyChangeData]: ...


class OccupancySensor(Sensor[OccupancySensorProperties, TStorage, str], Generic[TStorage]):
    """Occupancy sensor for detecting presence in a room or area."""

    _requires_frames = False

    def __init__(self, name: str = "Occupancy Sensor") -> None:
        super().__init__(name)
        self._write_state({OccupancyProperty.Detected.value: False})

    @property
    def type(self) -> SensorType:
        return SensorType.Occupancy

    @property
    def category(self) -> SensorCategory:
        return SensorCategory.Sensor

    @property
    def detected(self) -> bool:
        return bool(self.props.detected)

    def setDetected(self, value: bool) -> None:
        """Report occupancy state.

        Args:
            value: True when the area is currently occupied.

        Example:
            ```python
            occupancy.setDetected(True)
            ```
        """
        self._write_state({OccupancyProperty.Detected.value: value})

    async def updateValue(self, property: str, value: Any) -> None:
        """Read-only sensor: external writes are ignored."""
        # No-op — occupancy state is reported by the plugin, not set externally.
